include/nbl/builtin/glsl/ext/FFT/fft.glsl

```diff
@@ -66,258 +66,161 @@ vec2 nbl_glsl_ext_FFT_getData(in uvec3 coordinate, in uint channel);
 void nbl_glsl_ext_FFT_setData(in uvec3 coordinate, in uint channel, in vec2 complex_value);
 #endif
 
+#ifndef _NBL_GLSL_EXT_FFT_GET_PADDED_DATA_DECLARED_
+#define _NBL_GLSL_EXT_FFT_GET_PADDED_DATA_DECLARED_
+vec2 nbl_glsl_ext_FFT_getPaddedData(in uvec3 coordinate, in uint channel);
+#endif
+
 #ifndef _NBL_GLSL_EXT_FFT_GET_DATA_DEFINED_
 #error "You need to define `nbl_glsl_ext_FFT_getData` and mark `_NBL_GLSL_EXT_FFT_GET_DATA_DEFINED_`!"
 #endif
 #ifndef _NBL_GLSL_EXT_FFT_SET_DATA_DEFINED_
 #error "You need to define `nbl_glsl_ext_FFT_setData` and mark `_NBL_GLSL_EXT_FFT_SET_DATA_DEFINED_`!"
 #endif
+#ifndef _NBL_GLSL_EXT_FFT_GET_PADDED_DATA_DEFINED_
+#error "You need to define `nbl_glsl_ext_FFT_getPaddedData` and mark `_NBL_GLSL_EXT_FFT_GET_PADDED_DATA_DEFINED_`!"
+#endif
 
 // Count Leading Zeroes (naive?)
-uint clz(in uint x) 
+uint nbl_glsl_ext_FFT_clz(in uint x) 
 {
-    uint n = 0;
-	if (x == 0) { return 32; }
-    if (x <= 0x0000ffff) { n += 16; x <<= 16; }
-    if (x <= 0x00ffffff) { n +=  8; x <<= 8; }
-    if (x <= 0x0fffffff) { n +=  4; x <<= 4; }
-    if (x <= 0x3fffffff) { n +=  2; x <<= 2; }
-    if (x <= 0x7fffffff) { n++; };
-    return n;
+    return 31u - findMSB(x);
 }
 
-uint reverseBits(in uint x)
+uint nbl_glsl_ext_FFT_reverseBits(in uint x)
 {
-	uint count = 4 * 8 - 1;
-	uint reverse_num = x; 
-
-    x >>= 1;  
-    while(x > 0) 
-    { 
-       reverse_num <<= 1;        
-       reverse_num |= x & 1; 
-       x >>= 1; 
-       count--; 
-    } 
-    reverse_num <<= count;
-    return reverse_num; 
+    return bitfieldReverse(x);
 }
 
-uint calculate_twiddle_power(in uint threadId, in uint iteration, in uint logTwoN, in uint N) 
+uint nbl_glsl_ext_FFT_calculateTwiddlePower(in uint threadId, in uint iteration, in uint logTwoN) 
 {
-    return (threadId & ((N / (1u << (logTwoN - iteration))) * 2 - 1)) * ((1u << (logTwoN - iteration)) / 2);;
+    const uint shiftSuffix = logTwoN - 1u - iteration; // can we assert that iteration<logTwoN always?? yes
+    const uint suffixMask = (2u << iteration) - 1u;
+    return (threadId & suffixMask) << shiftSuffix;
 }
 
-vec2 twiddle(in uint threadId, in uint iteration, in uint logTwoN, in uint N) 
+nbl_glsl_complex nbl_glsl_ext_FFT_twiddle(in uint threadId, in uint iteration, in uint logTwoN, in uint N) 
 {
-    uint k = calculate_twiddle_power(threadId, iteration, logTwoN, N);
-    return nbl_glsl_eITheta(-1 * 2 * nbl_glsl_PI * k / N);
+    uint k = nbl_glsl_ext_FFT_calculateTwiddlePower(threadId, iteration, logTwoN);
+    return nbl_glsl_expImaginary(-1 * 2 * nbl_glsl_PI * k / N);
```

Comment:
make it clear what are types of the literal constants

also you're got `k/N` there which looks like integer division, its misleading because if you had integer division there, your code wouldn't work at all

Reply:
see my previous comment about not having implicit casts

Also you don't need `N` in the function arguments